library/Zend/Form/Form.php

```diff
@@ -807,8 +807,31 @@ public function attachInputFilterDefaults(InputFilterInterface $inputFilter, Fie
                     if ($childFieldset->getObject() instanceof InputFilterAwareInterface) {
                         $inputFilter->add($childFieldset->getObject()->getInputFilter(), $name);
                     } else {
-                        if ($fieldset instanceof Collection && $inputFilter instanceof CollectionInputFilter) {
-                            continue;
+                        // Add input filter for collections via getInputFilterSpecification()
+                        if ($childFieldset instanceof Collection
+                            && $childFieldset->getTargetElement() instanceof InputFilterProviderInterface
+                            && $childFieldset->getTargetElement()->getInputFilterSpecification()
+                        ) {
+                            $collectionContainerFilter = new CollectionInputFilter();
+
+                            $spec = $childFieldset->getTargetElement()->getInputFilterSpecification();
+                            $filter = $inputFactory->createInputFilter($spec);
+
+                            $collectionContainerFilter->setInputFilter($filter);
+
+                            $inputFilter->add($collectionContainerFilter, $name);
+
+                            // We need to copy the inputs to the collection input filter
+                            if ($inputFilter instanceof CollectionInputFilter) {
+                                foreach ($inputFilter->getInputs() as $filterName => $input) {
+                                    if (!$inputFilter->getInputFilter()->has($filterName)) {
+                                        $inputFilter->getInputFilter()->add($input, $filterName);
+                                    }
+                                }
+                            }
```

Comment:
I think it can be copied into separate private method to reduce repetitive addition block code